### ISSUE TITLE
Matcher error: received value must be a mock or spy function (when running interaction test in the console)

### ISSUE THREAD

**Issue by rexkenley** [storybookjs/storybook#16941]
**Describe the bug**
The interaction test works fine in the browser. 
![image](https://user-images.githubusercontent.com/10548251/145101961-4c5b1582-a503-4491-ad77-3d7fbba21492.png)

However, it doesn't work when run with npm t
![image](https://user-images.githubusercontent.com/10548251/145101550-aa9360db-0df4-4590-ac3c-37dc37b6cd4c.png)
![image](https://user-images.githubusercontent.com/10548251/145101639-14d43784-88d1-4205-b572-9c15f265f163.png)

**To Reproduce**
https://github.com/rexkenley/rollupError
npm t .\stories\components\USID\USID.spec.tsx

**System**
System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 96.0.4664.45
    Edge: Spartan (44.19041.1266.0), Chromium (96.0.1054.43)
  npmPackages:
    @storybook/addon-actions: ^6.4.8 => 6.4.8
    @storybook/addon-essentials: ^6.4.8 => 6.4.8
    @storybook/addon-interactions: ^6.4.8 => 6.4.8
    @storybook/addon-jest: ^6.4.8 => 6.4.8
    @storybook/addon-links: ^6.4.8 => 6.4.8
    @storybook/jest: ^0.0.5 => 0.0.5
    @storybook/react: ^6.4.8 => 6.4.8
    @storybook/testing-library: ^0.0.7 => 0.0.7
    @storybook/testing-react: ^1.2.2 => 1.2.2

**Comment by shilman** [comment]
@rexkenley This is probably an issue with `@storybook/testing-react`. When Storybook starts up, it runs a bunch of setup functions (called presets) from each of the addons that are installed. The `@storybook/addon-actions` preset automatically installs Jest spies on actions. However `@storybook/testing-react` doesn't run presets currently, hence the mismatch in behavior. @yannbf should be able to help with a workaround, and medium term we probably need to fix the issue in the `@storybook/testing-x` libraries.

**Comment by geiserman** [comment]
I noticed a similar error in the following scenario
when your interaction test is not in the first place order

example.stories.js
`export const DefaultExample = {
    args: { ...basicArgs },
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);
        await expect(canvas).not.toBeNull();
    },
};

export const DefaultExampleWithIncorrectEmail = {
    args: { ...basicArgs },
    play: async ({ args, canvasElement }) => {
        const canvas = within(canvasElement);
        await userEvent.type(canvas.getByPlaceholderText('Email Address *'), 'test');
        await userEvent.type(canvas.getByPlaceholderText('Password'), '12345678Z#');
        await userEvent.click(canvas.getByRole('button', { name: 'Log in', wait: 100 }));
        const emailErr = await screen.getByText('Email is invalid');
        await expect(emailErr).not.toBeNull();
        await expect(args.signIn).not.toHaveBeenCalled();
    },
};`

I am getting the error for DefaultExampleWithIncorrectEmail
![image](https://user-images.githubusercontent.com/44335369/148202942-7d1eb5b4-9201-4cef-9dbf-0baa7aeda04f.png)

`expect(signIn).not.toHaveBeenCalled()
expect(received).not.toHaveBeenCalled()

Matcher error: received value must be a mock or spy function

Received has type:  function
Received has value: "[Function anonymous]"`

For me, it only works if I change the order of the scenarios to:
`
example.stories.js

export const DefaultExampleWithIncorrectEmail = {
    args: { ...basicArgs },
    play: async ({ args, canvasElement }) => {
        const canvas = within(canvasElement);
        await userEvent.type(canvas.getByPlaceholderText('Email Address *'), 'test');
        await userEvent.type(canvas.getByPlaceholderText('Password'), '12345678Z#');
        await userEvent.click(canvas.getByRole('button', { name: 'Log in', wait: 100 }));
        const emailErr = await screen.getByText('Email is invalid');
        await expect(emailErr).not.toBeNull();
        await expect(args.signIn).not.toHaveBeenCalled();
    },
};

export const DefaultExample = {
    args: { ...basicArgs },
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);
        await expect(canvas).not.toBeNull();
    },
};
`
![image](https://user-images.githubusercontent.com/44335369/148202768-684ac01a-2580-4955-ae2e-c52eecc3c76b.png)

**Comment by shilman** [comment]
@geiserman i think this is a different error than the original issue. Do you a have a reproduction repo you can share? If not, can you create one? See [how to create a repro](https://storybook.js.org/docs/react/contribute/how-to-reproduce). Thank you! 🙏

**Comment by sambunting** [comment]
Is there any update to this issue? I'm having a similar issue with my tests.

The difference is - I'm not using @storybook/testing-react, I'm using just @storybook/testing-library at the moment.

I get the same error - but I only get it when it seems to be ran on a built Storybook? 
* When I run `npm run test-storybook` while Storybook is running on the local server, it's fine.
* When I run `npm run build-storybook`, use [serve](https://www.npmjs.com/package/serve) to host it, and then run `npm run test-storybook` I get the error.
* When I also run it in our CI [against a locally built version of Storybook](https://github.com/storybookjs/test-runner#2-running-against-locally-built-storybooks-in-ci) - I also get the error.

**Comment by yannbf** [comment]
> Is there any update to this issue? I'm having a similar issue with my tests.
> 
> The difference is - I'm not using @storybook/testing-react, I'm using just @storybook/testing-library at the moment.
> 
> I get the same error - but I only get it when it seems to be ran on a built Storybook?
> 
> * When I run `npm run test-storybook` while Storybook is running on the local server, it's fine.
> * When I run `npm run build-storybook`, use [serve](https://www.npmjs.com/package/serve) to host it, and then run `npm run test-storybook` I get the error.
> * When I also run it in our CI [against a locally built version of Storybook](https://github.com/storybookjs/test-runner#2-running-against-locally-built-storybooks-in-ci) - I also get the error.

Thanks for reporting this! Could you show us a repro so we can investigate the issue? Thanks!

**Comment by sambunting** [comment]
@yannbf I've created a repo here - hopefully it can be helpful! https://github.com/sambunting/storybook-jest-mock

**Comment by thomsocialbrothers** [comment]
Is there any update on this? I am experiencing the same issue when building storybook and then running the interaction tests. I am using Vite to build storybook.

**Comment by Nerlin** [comment]
I had the same problem today and I was able to fix it by removing a custom `config.optimization` entry from my `.storybook/main.js` file in the `webpackFinal` attribute:

```
-    config.optimization = {
-      splitChunks: {
-        chunks: "all",
-        minSize: 30 * 1024, // 30KB
-        maxSize: 1024 * 1024, // 1MB
-      }
-    };
```

**Comment by abhi40308** [comment]
i think updating storybook addons versions will fix this issue

**Comment by sambunting** [comment]
> i think updating storybook addons versions will fix this issue

I've ran `npm update` in the test repository I posted a while ago, then ran `npm run test-storybook:ci` and I seem to be getting a different error:

```
[TEST]  FAIL   browser: chromium  src/components/Button/Button.stories.jsx
[TEST]   ● Test suite failed to run
[TEST]
[TEST]     TypeError: (0 , _store.autoTitle) is not a function
[TEST]
[TEST]       at getDefaultTitle (node_modules/@storybook/test-runner/dist/cjs/playwright/transformPlaywright.js:72:31)
[TEST]
[TEST] Test Suites: 1 failed, 1 total
```

Updating just `@storybook/test-runner` and `@storybook/testing-library` to the latest versions (`0.7.1` and `0.0.13`) will give me the same `Matcher error: received value must be a mock or spy function` error.

Then updating `@storybook/builder-vite` and `vite` (A required dependency for the builder) to their latest versions (`0.2.2` and `3.1.1`) - running the `test-storybook:ci` command still gets me the error.

**Comment by Grsmto** [comment]
The way I have worked around that is by defining the args as a mock function directly:

```
import { jest } from '@storybook/jest';

MyComponent.args = {
  onChange: jest.fn(),
};

MyComponent.play = () => {
  ...
  expect(args.onChange).toHaveBeenCalled()
}
```

**Comment by chaosmirage** [comment]
> I get the same error - but I only get it when it seems to be ran on a built Storybook?

I have the same problem, found a cause, it is [here](https://github.com/storybookjs/storybook/blob/fbde656fb5c6fd0f1a1167144ce43102ba21248f/code/addons/interactions/src/preview.ts#LL43C1-L49C6)

```    
if (typeof val === 'function' && val.isAction) {
...
}
```

In built storybook version `isAction` will be something like `r` so the body of the conditional construct is ignored.
Inside conditional body is logic linked to mocks.

**Comment by kasir-barati** [comment]
Same issue here, had to use @Grsmto;s solution 😢 

Thanks @Grsmto ❤️ 

Any progress to solve this issue?

**Comment by shilman** [comment]
We've redone the way mocking is done in Storybook. For anybody encountering this issue, I'd recommend upgrading to the latest and filing a new issue if you run into any problems. It's available in 7.6 and should be a straightforward upgrade if you're already on 7.x.

More info: https://storybook.js.org/blog/storybook-test/

**Comment by yannbf** [comment]
Hey everyone! Please follow Michael's instructions above. The issue should be solved. If you do encounter futher issues, please report with a reproduction, and we'll be happy to take a look!
